src/sentry/api_gateway/proxy.py

```diff
@@ -0,0 +1,54 @@
+"""
+Utilities related to proxying a request to a region silo
+"""
+
+from urllib.parse import urljoin
+
+import requests as external_request
+from requests import Response as external_response
+from rest_framework.exceptions import NotFound
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from sentry.api.utils import generate_region_url
+from sentry.models.organization import Organization
+
+
+def get_org_region(slug: str) -> str:
+    from sentry.types.region import get_region_for_organization
+
+    org = Organization.objects.get(slug=slug)
+    region_info = get_region_for_organization(org)
+    return region_info.address
+
+
+def build_url(region: str, request: Request) -> str:
+    """Might need some changes to handle other format region address."""
+    region_url = generate_region_url(region)
+    return urljoin(region_url, request.path)
+
+
+def parse_response(response: external_response) -> Response:
+    """
+    Convert the Responses class from requests into the drf Response
+    """
+    return Response(data=response.json(), headers=response.headers, status=response.status_code)
+
+
+def proxy_request(request: Request, org_slug: str) -> Response:
+    """Take a django request opject and proxy it to a remote location given an org_slug"""
+
+    try:
+        region_location = get_org_region(org_slug)
+    except Organization.DoesNotExist:
+        raise NotFound(detail="Resource could not be found")
+    target_url = build_url(region_location, request)
+
+    query_params = getattr(request, request.method, None)
+    request_args = {
+        "headers": request.headers,
+        "params": dict(query_params) if query_params is not None else None,
+    }
+    resp: external_response = external_request.request(request.method, target_url, **request_args)
```

Comment:
It really makes me ponder if we could design this proxy more transparently, or even just some other proxy handling library that might exist.  Ideally, all we should be doing is potentially opening a socket and dumping the unparsed HTTP request through with the `Host` header changed?

Committing to parsing and repackaging the proxied request is a non trivial performance hit for every such request.

Reply:
When I was thinking about this earlier, I was wondering whether it would be better to forward the user's credentials or, if the original request has already been authenticated, to just use some kind of sudo authorization to make the proxy request.

A possible advantage to giving sudo authorization to the proxy request is that we could make multiple proxy requests through one persistent session (actually a pool of sessions, since we'd be handling many concurrently), which may perform better. I think https://stackoverflow.com/q/34837026 talks about roughly how we'd want to do that through the `requests` library. That's assuming it's even best to handle it in the app layer in the first place; I think there are also network-level approaches that are a bit over my head.

All of that, I'd be fine with breaking out as a later task to improve performance or scalability. @corps's question is more immediately relevant because it potentially simplifies the code you're working on right now.

Reply:
The libraries i see online for django proxying require a static proxy location. In our case the target server is dependent on the organization.

Reply:
did we look at something like https://docs.mitmproxy.org/stable/addons-examples/#http-stream-modify that has a python API for http proxying already implemented so we don't have to reinvent this? Even if we have to modify a library, that might be better than a less featureful/documented reimplementation